src/mlpack/methods/bayesian_linear_regression/bayesian_linear_regression.hpp

```diff
@@ -116,28 +121,154 @@ class BayesianLinearRegression
                            const size_t maxIterations = 50,
                            const double tolerance = 1e-4);
 
+  /**
+   * Create the BayesianLinearRegression object and train the model.  The
+   * regularization parameter is automatically set to its optimal value by
+   * maximization of the maginal likelihood.
+   *
+   * @param data Column-major input data, dim(P, N).
+   * @param responses A vector of targets, dim(N).
+   * @param centerData Whether or not center the data according to the
+   *    examples.
+   * @param scaleData Whether or not scale the data according to the
+   *    standard deviation of each feature.
+   * @param maxIterations Maximum number of iterations for convergency.
+   * @param tolerance Level from which the solution is considered sufficientlly
+   *    stable.
+   * @return Root mean squared error.
+   */
+  template<typename MatType,
+           typename ResponsesType,
+           typename = typename std::enable_if<
+               std::is_same<typename ResponsesType::elem_type, ElemType>::value
+           >::type>
+  BayesianLinearRegression(const MatType& data,
+                           const ResponsesType& responses,
+                           const bool centerData = true,
+                           const bool scaleData = false,
+                           const size_t maxIterations = 50,
+                           const double tolerance = 1e-4);
+
   /**
    * Run BayesianLinearRegression. The input matrix (like all mlpack matrices)
    * should be column-major -- each column is an observation and each row is a
    * dimension.
    *
    * @param data Column-major input data, dim(P, N).
    * @param responses A vector of targets, dim(N).
+   * @param centerData Whether or not center the data according to the
+   *    examples.
+   * @param scaleData Whether or not scale the data according to the
+   *    standard deviation of each feature.
+   * @param maxIterations Maximum number of iterations for convergency.
+   * @param tolerance Level from which the solution is considered sufficientlly
+   *    stable.
    * @return Root mean squared error.
    */
-  double Train(const arma::mat& data,
-               const arma::rowvec& responses);
+  // Many overloads necessary here until std::optional is available with C++17.
+  // The first overload is also necessary to avoid confusing the hyperparameter
+  // tuner, so that this can be correctly detected as a regression algorithm.
+  template<typename MatType>
+  ElemType Train(const MatType& data,
+                 const arma::rowvec& responses);
```

Comment:
Good to have the comment, I was wondering why and then I read it

Reply:
It took me hours to debug it the first time I encountered it... the algorithm was being detected as a classifier and it turns out the way to fix it was this extra overload.  I wonder if it might be better to have each classifier/regressor come with some traits that describe some things about it, but that can be a project for another time.